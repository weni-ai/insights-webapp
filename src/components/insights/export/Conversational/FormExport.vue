<template>
  <section
    class="conversational-export-form"
    data-testid="conversational-export-form"
  >
    <header>
      <p
        class="conversational-export-form__description"
        data-testid="conversational-export-form-description"
      >
        {{ $t('export_data.description') }}
      </p>
    </header>
    <section
      class="conversational-export-form__content"
      data-testid="conversational-export-form-content"
    >
      <ExportFilterDate
        v-model="date_range"
        data-testid="conversational-export-filter-date"
        :label="$t('export_data.select_data.label')"
        :placeholder="$t('export_data.select_period.placeholder')"
        :options="shortCutOptions"
        :minDate="getMinDate()"
        :maxDate="getMaxDate()"
        @update:model-value="updateDateRange"
        @select-date="updateSelectDateRange"
      />

      <FormCheckbox data-testid="conversational-form-checkbox-component" />

      <ExportFooter
        data-testid="conversational-export-footer"
        :selectedFormat="selectedFormat"
        :acceptTerms="accept_terms"
        :formatLabel="$t('export_data.select_format')"
        :warningTermsText="$t('export_data.warning_terms')"
        :acceptTermsText="$t('export_data.accept_terms')"
        @format-change="updateFormat"
        @accept-terms-change="updateAcceptTerms"
      />
    </section>
  </section>
</template>

<script setup lang="ts">
import ExportFilterDate from '../ExportFilterDate.vue';
import ExportFooter from '../ExportFooter.vue';
import FormCheckbox from './FormCheckbox.vue';
import { useConversationalExport } from '@/store/modules/export/conversational/export';
import { storeToRefs } from 'pinia';
import { computed, ref } from 'vue';
import { useI18n } from 'vue-i18n';
import { format, subDays, addDays, isValid, parseISO, isAfter } from 'date-fns';

const conversationalExport = useConversationalExport();
const { setDateRange, setType, setAcceptTerms } = conversationalExport;
const { date_range, type, accept_terms } = storeToRefs(conversationalExport);

const { t } = useI18n();

const selectedFormat = computed(() => {
  return type.value === 'CSV' ? '.csv' : '.xlsx';
});

const selectDateRange = ref({ start: '', end: '' });

const shortCutOptions = computed(() => [
  {
    name: t('export_data.select_data.shortcuts.last_7_days'),
    id: 'last-7-days',
  },
  {
    name: t('export_data.select_data.shortcuts.last_14_days'),
    id: 'last-14-days',
  },
  {
    name: t('export_data.select_data.shortcuts.last_30_days'),
    id: 'last-30-days',
  },
  {
    name: t('export_data.select_data.shortcuts.last_60_days'),
    id: 'last-60-days',
  },
  {
    name: t('export_data.select_data.shortcuts.last_90_days'),
    id: 'last-90-days',
  },
  {
    name: t('export_data.select_data.shortcuts.current_month'),
    id: 'current-month',
  },
  {
    name: t('export_data.select_data.shortcuts.previous_month'),
    id: 'previous-month',
  },
]);

const getMinDate = (): string => {
  const currentSelection = selectDateRange.value;

  const defaultMin = format(subDays(new Date(), 92), 'yyyy-MM-dd');

  if (currentSelection.start === '' && currentSelection.end === '') {
    return null;
  }

  if (!currentSelection || !currentSelection.start) {
    return defaultMin;
  }

  const startDate = parseISO(currentSelection.start);

  if (
    isValid(startDate) &&
    (!currentSelection.end || currentSelection.start === currentSelection.end)
  ) {
    const calculatedMin = format(subDays(startDate, 92), 'yyyy-MM-dd');
    const calculatedMinDate = parseISO(calculatedMin);
    return isValid(calculatedMinDate) ? calculatedMin : defaultMin;
  }

  if (!isValid(startDate)) {
    return null;
  }

  if (isValid(startDate) && defaultMin !== format(startDate, 'yyyy-MM-dd')) {
    return format(subDays(startDate, 92), 'yyyy-MM-dd');
  }

  return defaultMin;
};

const getMaxDate = (): string => {
  const today = new Date();
  const currentSelection = selectDateRange.value;
  const defaultMax = format(today, 'yyyy-MM-dd');

  if (!currentSelection || !currentSelection.start) {
    return defaultMax;
  }

  const startDate = parseISO(currentSelection.start);

  if (
    isValid(startDate) &&
    (!currentSelection.end || currentSelection.start === currentSelection.end)
  ) {
    const calculatedMax = addDays(startDate, 92);
    if (isAfter(calculatedMax, today)) {
      return defaultMax;
    }
    return format(calculatedMax, 'yyyy-MM-dd');
  }

  return defaultMax;
};

const updateDateRange = (value: { start: string; end: string }) => {
  setDateRange(value.start, value.end);
};

const updateSelectDateRange = (value: { start: string; end: string }) => {
  selectDateRange.value = value;
};

const updateFormat = (format: '.csv' | '.xlsx') => {
  setType(format === '.csv' ? 'CSV' : 'XLSX');
};

const updateAcceptTerms = (value: boolean) => {
  setAcceptTerms(value);
};
</script>

<style scoped lang="scss">
.conversational-export-form {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-sm;
  background: $unnnic-color-neutral-white;

  &__content {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-sm;
    :deep(.unnnic-label__label) {
      margin: 0;
    }
  }

  &__description {
    color: $unnnic-color-neutral-cloudy;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
