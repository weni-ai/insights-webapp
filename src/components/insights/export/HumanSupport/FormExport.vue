<template>
  <section
    class="export-data-form"
    data-testid="export-data-form"
  >
    <header>
      <p
        class="export-data-form__description"
        data-testid="export-data-form-description"
      >
        {{ $t('export_data.description') }}
      </p>
    </header>
    <section
      class="export-data-form__content"
      data-testid="export-data-form-content"
    >
      <ExportFilterDate
        v-model="date_range"
        data-testid="export-data-filter-date"
        :label="$t('export_data.select_data.label')"
        :placeholder="$t('export_data.select_data.placeholder')"
        :options="shortCutOptions"
        :minDate="getMinDate()"
        :maxDate="getMaxDate()"
        @update:model-value="updateDateRange"
        @select-date="updateSelectDateRange"
      />

      <section
        class="export-data-form__chats-status"
        data-testid="export-data-form-chats-status"
      >
        <UnnnicRadio
          :data-testid="'radio-chats-open'"
          :modelValue="selectedChatStatus"
          value="open"
          @update:model-value="updateChatStatus('open')"
        >
          {{ $t('export_data.chats_open') }}
        </UnnnicRadio>
        <UnnnicRadio
          :data-testid="'radio-chats-closed'"
          :modelValue="selectedChatStatus"
          value="closed"
          @update:model-value="updateChatStatus('closed')"
        >
          {{ $t('export_data.chats_closed') }}
        </UnnnicRadio>
      </section>

      <section
        class="export-data-form__filters"
        data-testid="export-data-form-filters"
      >
        <section class="export-data-form__filters-container">
          <UnnnicLabel :label="$t('export_data.filters.sector')" />
          <FilterMultiSelect
            v-model="sectors"
            :placeholder="$t('export_data.filters.select_sector')"
            source="sectors"
            keyValueField="uuid"
            :allLabel="$t('export_data.filters.all_sectors')"
            @update:model-value="updateSectors"
          />
        </section>

        <section class="export-data-form__filters-container">
          <UnnnicLabel :label="$t('export_data.filters.queue')" />
          <FilterMultiSelect
            v-model="queues"
            :placeholder="$t('export_data.filters.select_queue')"
            source="queues"
            keyValueField="uuid"
            :allLabel="$t('export_data.filters.all_queues')"
            :disabled="!hasSectorsSelected || isManySectorsSelected"
            :dependsOnValue="dependsOnValueQueues"
            @update:model-value="updateQueues"
          />
        </section>

        <section class="export-data-form__filters-container">
          <UnnnicLabel :label="$t('export_data.filters.agent')" />
          <FilterMultiSelect
            v-model="agents"
            :placeholder="$t('export_data.filters.select_agent')"
            source="agents"
            keyValueField="uuid"
            :allLabel="$t('export_data.filters.all_agents')"
            :disabled="!hasSectorsSelected"
            :dependsOnValue="dependsOnValueAgents"
            @update:model-value="updateAgents"
          />
        </section>

        <section class="export-data-form__filters-container">
          <UnnnicLabel :label="$t('export_data.filters.tag')" />
          <FilterMultiSelect
            v-model="tags"
            :placeholder="$t('export_data.filters.select_tag')"
            source="tags"
            keyValueField="uuid"
            :allLabel="$t('export_data.filters.all_tags')"
            :disabled="!hasSectorsSelected || isManySectorsSelected"
            :dependsOnValue="dependsOnValueTags"
            @update:model-value="updateTags"
          />
        </section>
      </section>

      <FormCheckbox data-testid="human-support-form-checkbox-component" />

      <ExportFooter
        data-testid="human-support-export-footer"
        :selectedFormat="selectedFormat"
        :acceptTerms="accept_terms"
        :formatLabel="$t('export_data.select_format')"
        :warningTermsText="$t('export_data.warning_terms')"
        :acceptTermsText="$t('export_data.accept_terms')"
        @format-change="updateFormat"
        @accept-terms-change="updateAcceptTerms"
      />
    </section>
  </section>
</template>

<script setup lang="ts">
import FilterMultiSelect from '@/components/insights/Layout/HeaderFilters/FilterMultiSelect.vue';
import FormCheckbox from './FormCheckbox.vue';
import ExportFooter from '../ExportFooter.vue';
import ExportFilterDate from '../ExportFilterDate.vue';
import { useHumanSupportExport } from '@/store/modules/export/humanSupport/export';
import { storeToRefs } from 'pinia';
import { computed, ref } from 'vue';
import { useI18n } from 'vue-i18n';
import { format, subDays, addDays, isValid, parseISO, isAfter } from 'date-fns';

const humanSupportExport = useHumanSupportExport();
const {
  setStatusChats,
  setDateRange,
  setSectors,
  setAgents,
  setQueues,
  setTags,
  setType,
  setAcceptTerms,
} = humanSupportExport;
const {
  open_chats,
  sectors,
  agents,
  queues,
  tags,
  type,
  accept_terms,
  date_range,
} = storeToRefs(humanSupportExport);

const selectedFormat = computed(() => {
  return type.value === '.csv' ? '.csv' : '.xlsx';
});

const selectedChatStatus = computed(() => {
  return open_chats.value ? 'open' : 'closed';
});

const hasSectorsSelected = computed(() => {
  return sectors.value && sectors.value.length > 0;
});

const sectorsForDependency = computed(() => {
  return sectors.value?.map((sector: any) => sector.value).join(',') || '';
});

const isManySectorsSelected = computed(() => {
  return sectors.value?.length > 1 && sectorsForDependency.value !== '__all__';
});

const dependsOnValueQueues = computed(() => {
  if (sectors.value?.length === 1 && sectorsForDependency.value !== '__all__') {
    return { sector_id: sectorsForDependency.value };
  }
  return { sectors: sectorsForDependency.value };
});

const dependsOnValueAgents = computed(() => {
  if (sectors.value?.length === 1 && sectorsForDependency.value !== '__all__') {
    return { sector_id: sectorsForDependency.value };
  }
  return { sectors: sectorsForDependency.value };
});

const dependsOnValueTags = computed(() => {
  if (sectors.value?.length === 1 && sectorsForDependency.value !== '__all__') {
    return { sector_id: sectorsForDependency.value };
  }
  return { sectors: sectorsForDependency.value };
});

const updateDateRange = (value: { start: string; end: string }) => {
  setDateRange(value.start, value.end);
};

const updateSelectDateRange = (value: { start: string; end: string }) => {
  selectDateRange.value = value;
};

const updateChatStatus = (status: string) => {
  setStatusChats(status === 'open');
};

const updateSectors = (value: any[]) => {
  setSectors(value);
  setAgents([]);
  setQueues([]);
  setTags([]);
};

const updateAgents = (value: any[]) => {
  setAgents(value);
};

const updateQueues = (value: any[]) => {
  setQueues(value);
};

const updateTags = (value: any[]) => {
  setTags(value);
};

const updateFormat = (format: '.csv' | '.xlsx') => {
  setType(format);
};

const updateAcceptTerms = (value: boolean) => {
  setAcceptTerms(value);
};

const { t } = useI18n();

const selectDateRange = ref({ start: '', end: '' });

const shortCutOptions = computed(() => [
  {
    name: t('export_data.select_data.shortcuts.last_7_days'),
    id: 'last-7-days',
  },
  {
    name: t('export_data.select_data.shortcuts.last_14_days'),
    id: 'last-14-days',
  },
  {
    name: t('export_data.select_data.shortcuts.last_30_days'),
    id: 'last-30-days',
  },
  {
    name: t('export_data.select_data.shortcuts.last_60_days'),
    id: 'last-60-days',
  },
  {
    name: t('export_data.select_data.shortcuts.last_90_days'),
    id: 'last-90-days',
  },
  {
    name: t('export_data.select_data.shortcuts.current_month'),
    id: 'current-month',
  },
  {
    name: t('export_data.select_data.shortcuts.previous_month'),
    id: 'previous-month',
  },
]);

const getMinDate = (): string => {
  const currentSelection = selectDateRange.value;

  const defaultMin = format(subDays(new Date(), 92), 'yyyy-MM-dd');

  if (currentSelection.start === '' && currentSelection.end === '') {
    return null;
  }

  if (!currentSelection || !currentSelection.start) {
    return defaultMin;
  }

  const startDate = parseISO(currentSelection.start);

  if (
    isValid(startDate) &&
    (!currentSelection.end || currentSelection.start === currentSelection.end)
  ) {
    const calculatedMin = format(subDays(startDate, 92), 'yyyy-MM-dd');
    const calculatedMinDate = parseISO(calculatedMin);
    return isValid(calculatedMinDate) ? calculatedMin : defaultMin;
  }

  if (!isValid(startDate)) {
    return null;
  }

  if (isValid(startDate) && defaultMin !== format(startDate, 'yyyy-MM-dd')) {
    return format(subDays(startDate, 92), 'yyyy-MM-dd');
  }

  return defaultMin;
};

const getMaxDate = (): string => {
  const today = new Date();
  const currentSelection = selectDateRange.value;
  const defaultMax = format(today, 'yyyy-MM-dd');

  if (!currentSelection || !currentSelection.start) {
    return defaultMax;
  }

  const startDate = parseISO(currentSelection.start);

  if (
    isValid(startDate) &&
    (!currentSelection.end || currentSelection.start === currentSelection.end)
  ) {
    const calculatedMax = addDays(startDate, 92);
    if (isAfter(calculatedMax, today)) {
      return defaultMax;
    }
    return format(calculatedMax, 'yyyy-MM-dd');
  }

  return defaultMax;
};
</script>

<style scoped lang="scss">
.export-data-form {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-sm;
  background: $unnnic-color-neutral-white;

  &__content {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-sm;
    :deep(.unnnic-label__label) {
      margin: 0;
    }
  }

  &__description {
    color: $unnnic-color-neutral-cloudy;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__filters-container {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-nano;
    flex: 1 1 calc(50% - #{$unnnic-spacing-sm / 2});
    min-width: 0;
  }

  &__filters {
    display: flex;
    flex-wrap: wrap;
    gap: $unnnic-spacing-sm;
  }

  &__chats-status {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-sm;
  }
}
</style>
