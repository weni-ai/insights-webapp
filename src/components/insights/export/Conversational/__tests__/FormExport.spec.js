import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import FormExport from '../FormExport.vue';

const mockStore = {
  date_range: { value: { start: '', end: '' } },
  type: { value: 'XLSX' },
  accept_terms: { value: false },
  setDateRange: vi.fn(),
  setType: vi.fn(),
  setAcceptTerms: vi.fn(),
};

vi.mock('@/store/modules/export/conversational/export', () => ({
  useConversationalExport: () => mockStore,
}));

vi.mock('pinia', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    storeToRefs: (store) => store,
  };
});

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: {
    en: {
      export_data: {
        description: 'Export description',
        select_data: {
          label: 'Date',
          shortcuts: {
            last_7_days: 'Last 7 days',
            last_14_days: 'Last 14 days',
            last_30_days: 'Last 30 days',
            last_60_days: 'Last 60 days',
            last_90_days: 'Last 90 days',
            current_month: 'Current month',
            previous_month: 'Previous month',
          },
        },
        select_period: { placeholder: 'Select' },
        select_format: 'Format',
        warning_terms: 'Warning',
        accept_terms: 'Accept',
      },
    },
  },
});

config.global.plugins = [i18n];

describe('Conversational FormExport', () => {
  let wrapper;

  const createWrapper = (storeOverrides = {}) => {
    Object.assign(mockStore, storeOverrides);

    return mount(FormExport, {
      global: {
        stubs: {
          ExportFilterDate: true,
          FormCheckbox: true,
          ExportFooter: true,
        },
      },
    });
  };

  const section = () =>
    wrapper.find('[data-testid="conversational-export-form"]');

  beforeEach(() => {
    vi.clearAllMocks();

    Object.assign(mockStore, {
      date_range: { value: { start: '', end: '' } },
      type: { value: 'XLSX' },
      accept_terms: { value: false },
    });

    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render main section', () => {
      expect(section().exists()).toBe(true);
    });

    it('should render description', () => {
      const description = wrapper.find(
        '[data-testid="conversational-export-form-description"]',
      );
      expect(description.exists()).toBe(true);
    });

    it('should render all child components', () => {
      expect(
        wrapper
          .find('[data-testid="conversational-export-filter-date"]')
          .exists(),
      ).toBe(true);
      expect(
        wrapper
          .find('[data-testid="conversational-form-checkbox-component"]')
          .exists(),
      ).toBe(true);
      expect(
        wrapper.find('[data-testid="conversational-export-footer"]').exists(),
      ).toBe(true);
    });
  });

  describe('Computed properties', () => {
    it('should compute selectedFormat as .xlsx when type is XLSX', () => {
      expect(wrapper.vm.selectedFormat).toBe('.xlsx');
    });

    it('should compute selectedFormat as .csv when type is CSV', () => {
      wrapper = createWrapper({ type: { value: 'CSV' } });
      expect(wrapper.vm.selectedFormat).toBe('.csv');
    });

    it('should compute shortCutOptions with correct translations', () => {
      const options = wrapper.vm.shortCutOptions;
      expect(options).toHaveLength(7);
      expect(options[0].id).toBe('last-7-days');
      expect(options[6].id).toBe('previous-month');
    });
  });

  describe('Date range methods', () => {
    it('should call getMinDate and return null for empty range', () => {
      const minDate = wrapper.vm.getMinDate();
      expect(minDate).toBeNull();
    });

    it('should call getMaxDate and return today for empty range', () => {
      const maxDate = wrapper.vm.getMaxDate();
      expect(maxDate).toBeTruthy();
    });

    it('should calculate minDate based on selectDateRange', async () => {
      wrapper.vm.selectDateRange = {
        start: '2024-01-01',
        end: '',
      };
      const minDate = wrapper.vm.getMinDate();
      expect(minDate).toBeTruthy();
    });

    it('should calculate maxDate based on selectDateRange', async () => {
      wrapper.vm.selectDateRange = {
        start: '2024-01-01',
        end: '',
      };
      const maxDate = wrapper.vm.getMaxDate();
      expect(maxDate).toBeTruthy();
    });
  });

  describe('Event handlers', () => {
    it('should update date range', () => {
      const dateRange = { start: '2024-01-01', end: '2024-01-31' };
      wrapper.vm.updateDateRange(dateRange);

      expect(mockStore.setDateRange).toHaveBeenCalledWith(
        '2024-01-01',
        '2024-01-31',
      );
    });

    it('should update select date range', () => {
      const dateRange = { start: '2024-01-01', end: '2024-01-31' };
      wrapper.vm.updateSelectDateRange(dateRange);

      expect(wrapper.vm.selectDateRange).toEqual(dateRange);
    });

    it('should update format to CSV', () => {
      wrapper.vm.updateFormat('.csv');
      expect(mockStore.setType).toHaveBeenCalledWith('CSV');
    });

    it('should update format to XLSX', () => {
      wrapper.vm.updateFormat('.xlsx');
      expect(mockStore.setType).toHaveBeenCalledWith('XLSX');
    });

    it('should update accept terms', () => {
      wrapper.vm.updateAcceptTerms(true);
      expect(mockStore.setAcceptTerms).toHaveBeenCalledWith(true);
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(section().classes()).toContain('conversational-export-form');
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
