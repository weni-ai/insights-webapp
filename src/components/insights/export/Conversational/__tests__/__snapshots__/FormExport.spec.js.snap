// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Conversational FormExport > Component structure > should match snapshot 1`] = `
<section
  class="conversational-export-form"
  data-testid="conversational-export-form"
  data-v-0721d1b7=""
>
  <header
    data-v-0721d1b7=""
  >
    <p
      class="conversational-export-form__description"
      data-testid="conversational-export-form-description"
      data-v-0721d1b7=""
    >
      export_data.description
    </p>
  </header>
  <section
    class="conversational-export-form__content"
    data-testid="conversational-export-form-content"
    data-v-0721d1b7=""
  >
    <export-filter-date-stub
      data-testid="conversational-export-filter-date"
      data-v-0721d1b7=""
      label="export_data.select_data.label"
      maxdate="2025-10-29"
      modelvalue="[object Object]"
      options="[object Object]"
      placeholder="export_data.select_period.placeholder"
    />
    <form-checkbox-stub
      data-testid="conversational-form-checkbox-component"
      data-v-0721d1b7=""
    />
    <export-footer-stub
      acceptterms="[object Object]"
      accepttermstext="export_data.accept_terms"
      data-testid="conversational-export-footer"
      data-v-0721d1b7=""
      formatlabel="export_data.select_format"
      selectedformat=".xlsx"
      warningtermstext="export_data.warning_terms"
    />
  </section>
</section>
`;
